src/concurrency/thread.rs

```diff
@@ -656,11 +648,18 @@ impl<'tcx> ThreadManager<'tcx> {
                     @capture<'tcx> {
                         joined_thread_id: ThreadId,
                     }
-                    @unblock = |this| {
-                        if let Some(data_race) = &mut this.machine.data_race {
-                            data_race.thread_joined(&this.machine.threads, joined_thread_id);
+                    @unblock = |this, tcb_state| {
+                        match tcb_state {
+                            MachineCallbackState::Ready => {
+                                if let Some(data_race) = &mut this.machine.data_race {
+                                    data_race.thread_joined(&this.machine.threads, joined_thread_id);
+                                }
+                                interp_ok(())
+                            }
+                            MachineCallbackState::TimedOut => {
+                                panic!("Join thread operation received unexpected timeout state - operations do not support timeouts")
```

Comment:
No need to write such a detailed error each time. Also, please avoid the rightwards drift. So instead of a `match unblock`, just have something like
```rust
assert_eq!(unblock, UnblockKind::Ready);
```
at the beginning of the function, and then go on with the function body as before.